File: src/main/java/hdzi/editstarters/buildsystem/ProjectFile.java
package hdzi.editstarters.buildsystem;

import com.intellij.psi.PsiElement;
import hdzi.editstarters.EditStarters;
import hdzi.editstarters.dependency.Bom;
import hdzi.editstarters.dependency.Point;
import hdzi.editstarters.dependency.Repository;
import hdzi.editstarters.dependency.StarterInfo;
import hdzi.editstarters.ui.ShowErrorException;
import org.apache.commons.collections.CollectionUtils;

import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Set;
import java.util.stream.Collectors;

public abstract class ProjectFile<T extends PsiElement> implements EditStarters {


    @Override
    public void removeStarters(Collection<StarterInfo> dependencies) {
        try {
            T dependenciesTag = getOrCreateDependenciesTag();
            // 取已存在的依赖
            List<ProjectDependency> extDependencies = findAllDependencies(dependenciesTag);
            // 转化待删除的依赖成字符串形式，方便对比
            Set<String> removeDependencies = dependencies.stream().map(Point::point).collect(Collectors.toSet());
            // 遍历存在的依赖，如果待删除的依赖包含它，就删除
            for (ProjectDependency extDependency : extDependencies) {
                if (removeDependencies.contains(extDependency.point())) {
                    PsiElement element = extDependency.getElement();
                    if (element != null) {
                        element.delete();
                    }
                }
            }
        } catch (ShowErrorException e) {
            throw e;
        } catch (Exception e) {
            throw new ShowErrorException("Syntax error!", e);
        }

    }

    @Override
    public void addStarters(Collection<StarterInfo> dependencies) {
        try {
            T dependenciesTag = getOrCreateDependenciesTag();
            for (StarterInfo dependency : dependencies) {
                createDependencyTag(dependenciesTag, dependency);
                Bom bom = dependency.getBom();
                if (bom != null) {
                    addBom(bom);
                }

                List<Repository> repositories = dependency.getRepositories();
                if (CollectionUtils.isNotEmpty(repositories)) {
                    addRepositories(repositories);
                }
            }
        } catch (ShowErrorException e) {
            throw e;
        } catch (Exception e) {
            throw new ShowErrorException("Syntax error!", e);
        }
    }

    /**
     * 添加bom信息
     */
    private void addBom(Bom bom) {
        T bomTag = getOrCreateBomsTag();
        // 去重后新建
        List<ProjectBom> allBoms = findAllBoms(bomTag);
        if (allBoms.stream().noneMatch(b -> Objects.equals(b.point(), bom.point()))) {
            createBomTag(bomTag, bom);
        }
    }

    /**
     * 添加仓库信息
     */
    private void addRepositories(Collection<Repository> repositories) {
        T repositoriesTag = getOrCreateRepositoriesTag();
        List<ProjectRepository> existingRepos = findAllRepositories(repositoriesTag);
        Set<String> existingRepoPointSet = existingRepos.stream().map(Point::point).collect(Collectors.toSet());
        for (Repository repository : repositories) {
            if (!existingRepoPointSet.contains(repository.point())) {
                createRepositoryTag(repositoriesTag, repository);
            }
        }
    }

    protected abstract T getOrCreateDependenciesTag();

    protected abstract List<ProjectDependency> findAllDependencies(T dependenciesTag);

    protected abstract void createDependencyTag(T dependenciesTag, StarterInfo info);

    protected abstract T getOrCreateBomsTag();

    protected abstract List<ProjectBom> findAllBoms(T bomsTag);

    protected abstract void createBomTag(T bomsTag, Bom bom);

    protected abstract T getOrCreateRepositoriesTag();

    protected abstract List<ProjectRepository> findAllRepositories(T repositoriesTag);

    protected abstract void createRepositoryTag(T repositoriesTag, Repository repository);
}

File: src/main/kotlin/hdzi/editstarters/ProjectFile.kt
package hdzi.editstarters

import com.intellij.psi.PsiElement
import hdzi.editstarters.bean.StarterInfo
import hdzi.editstarters.bean.initializr.InitializrBom
import hdzi.editstarters.bean.initializr.InitializrRepository
import hdzi.editstarters.bean.project.ProjectBom
import hdzi.editstarters.bean.project.ProjectDependency
import hdzi.editstarters.bean.project.ProjectRepository

abstract class ProjectFile<T : PsiElement> : EditStarters {

    override fun removeStarters(dependencies: Collection<StarterInfo>) {
        val dependenciesTag = getOrCreateDependenciesTag()
        // 取已存在的依赖
        val extDependencies = findAllDependencies(dependenciesTag)
        // 转化待删除的依赖成字符串形式，方便对比
        val removeDependencies = dependencies.map { it.point }.toSet()
        // 遍历存在的依赖，如果待删除的依赖包含它，就删除
        for (extDependency in extDependencies) {
            if (removeDependencies.contains(extDependency.point)) {
                extDependency.element!!.delete()
            }
        }
    }

    override fun addStarters(dependencies: Collection<StarterInfo>) {
        val dependenciesTag = getOrCreateDependenciesTag()

        dependencies.forEach {
            createDependencyTag(dependenciesTag, it)
            if (it.bom != null) addBom(it.bom!!)
            if (!it.repositories.isEmpty()) addRepositories(it.repositories)
        }
    }

    /**
     * 添加bom信息
     */
    private fun addBom(bom: InitializrBom) {
        val bomTag = getOrCreateBomsTag()
        // 去重后新建
        findAllBoms(bomTag).find { bom.point == it.point } ?: createBomTag(bomTag, bom)
    }

    /**
     * 添加仓库信息
     */
    private fun addRepositories(repositories: Set<InitializrRepository>) {
        val repositoriesTag = getOrCreateRepositoriesTag()
        val existingRepos = findAllRepositories(repositoriesTag).map { it.point }.toSet()
        repositories.forEach {
            if (!existingRepos.contains(it.point)) createRepositoryTag(repositoriesTag, it)
        }
    }

    protected abstract fun getOrCreateDependenciesTag(): T

    protected abstract fun findAllDependencies(dependenciesTag: T): Sequence<ProjectDependency>

    protected abstract fun createDependencyTag(dependenciesTag: T, info: StarterInfo)

    protected abstract fun getOrCreateBomsTag(): T

    protected abstract fun findAllBoms(bomsTag: T): Sequence<ProjectBom>

    protected abstract fun createBomTag(bomsTag: T, bom: InitializrBom)

    protected abstract fun getOrCreateRepositoriesTag(): T

    protected abstract fun findAllRepositories(repositoriesTag: T): Sequence<ProjectRepository>

    protected abstract fun createRepositoryTag(repositoriesTag: T, repository: InitializrRepository)
}
